### PR TITLE
Make ping unidirectional

### LINKED ISSUE #378: Make ping unidirectional
So that only the side which opened the substream can send pings.

Right now we track the pings we send, and anything we receive that doesn't match what we sent earlier is guessed as being a ping from the remote.
But it makes more sense if the two sides each open their own substream.

### PR DESCRIPTION
**Pull request by tomaka**
Fix #378

### CODE REVIEW

File: libp2p/examples/ping-client.rs [hunk 1]
```diff
@@ -78,15 +78,19 @@ fn main() {
...
+                let tx = tx.take();
+                let ping = pinger.ping().map_err(|_| unreachable!()).inspect(move |_| {
+                    println!("Received pong from the remote");
+                    if let Some(tx) = tx {
+                        let _ = tx.send(());
+                    }
+                });
+                ping.select(processing).map(|_| ()).map_err(|(e, _)| e)
+            } else {
+                unreachable!()
```
Comment (rphmeier) [inline libp2p/examples/ping-client.rs]: why is it unreachable? put a proof

File: libp2p/examples/ping-client.rs [hunk 2]
```diff
@@ -78,15 +78,19 @@ fn main() {
...
-                if let Some(tx) = tx {
-                    let _ = tx.send(());
-                }
-            });
-            ping.select(future).map(|_| ()).map_err(|(e, _)| e)
+        |out, _client_addr| {
+            if let libp2p::ping::PingOutput::Pinger { mut pinger, processing } = out {
+                let tx = tx.take();
+                let ping = pinger.ping().map_err(|_| unreachable!()).inspect(move |_| {
+                    println!("Received pong from the remote");
```
Comment (gnunicorn) [inline libp2p/examples/ping-client.rs]: `println`

File: ping/src/lib.rs [hunk 3]
```diff
@@ -130,86 +148,119 @@ where
...
+    // Ignore the errors if `tx` closed.
+    let rx = rx.then(|r| Ok(r.ok())).filter_map(|a| a);
+
+    let pinger = Pinger {
+        send: tx,
+        rng: EntropyRng::default(),
+    };
+
+    // Hashmap that associates outgoing payloads to one-shot senders.
+    // TODO: can't figure out how to make it work without using an Arc/Mutex
```
Comment (gnunicorn) [inline ping/src/lib.rs]: Are pongs not expected to be received in the same order as the pings were send? We aren't checking for that at the moment (they can respond in whichever order they want), but if that was implied by the protocol, we could simplify this codebase and use a FIFO-Buffer of expected pongs and have a `futures` per ping interface instead. However, not necessarily part of this PR as it is only touching the code base but concerned with other aspects of the protocol.

**Review by gnunicorn — APPROVED** [review]
LGTM, just remove the `println` ;) .

**Pull request merged**